ert/data/record/_record.py

```diff
@@ -0,0 +1,224 @@
+import pathlib
+from abc import ABC
+from collections import deque
+from enum import Enum
+from typing import Dict, List, Mapping, MutableMapping, Optional, Tuple, Union, cast
+
+from beartype import beartype
+from beartype.roar import BeartypeException  # type: ignore
+from pydantic import PositiveInt
+from ert.serialization import get_serializer
+
+number = Union[int, float]
+numerical_record_data = Union[
+    List[number],
+    Dict[str, number],
+    Dict[int, number],
+]
+blob_record_data = bytes
+record_data = Union[numerical_record_data, blob_record_data]
+record_collection = Tuple["Record", ...]
+RecordIndex = Tuple[Union[int, str], ...]
+
+
+def _build_record_index(
+    data: numerical_record_data,
+) -> RecordIndex:
+    if isinstance(data, MutableMapping):
+        return tuple(data.keys())
+    else:
+        return tuple(range(len(data)))
+
+
+class RecordValidationError(Exception):
+    pass
+
+
+class RecordType(str, Enum):
+    LIST_FLOAT = "LIST_FLOAT"
+    MAPPING_INT_FLOAT = "MAPPING_INT_FLOAT"
+    MAPPING_STR_FLOAT = "MAPPING_STR_FLOAT"
+    BYTES = "BYTES"
+
+
+class Record(ABC):
+    @property
+    def data(self) -> record_data:
+        pass
+
+    @property
+    def record_type(self) -> RecordType:
+        pass
+
+
+class BlobRecord(Record):
+    def __init__(self, data: blob_record_data) -> None:
+        self._record_type = RecordType.BYTES
+
+        try:
+            self._data = self._validate_data(data)
+        except BeartypeException as e:
+            raise RecordValidationError(str(e))
+
+    @beartype
+    def _validate_data(self, data: blob_record_data) -> blob_record_data:
+        return data
+
+    @property
+    def data(self) -> blob_record_data:
+        return self._data
+
+    @property
+    def record_type(self) -> RecordType:
+        return RecordType.BYTES
+
+    def __eq__(self, o: object) -> bool:
+        if isinstance(o, type(self)):
+            return self.__dict__ == o.__dict__
+        return False
+
+
+class NumericalRecord(Record):
+    def __init__(
+        self, data: numerical_record_data, index: Optional[RecordIndex] = None
+    ) -> None:
+        if isinstance(data, (set, frozenset, deque, tuple)):
+            data = [val for _, val in enumerate(data)]
+
+        try:
+            self._validate_data(data)
+        except BeartypeException as e:
+            raise RecordValidationError(str(e))
+        self._data = data
+
+        if index is None:
+            index = _build_record_index(data)
+        self._validate_index(data, index)
+        self._index = index
+
+        self._type = self._infer_type(data)
+
+    @beartype
+    def _validate_data(self, data: numerical_record_data) -> None:
+        # beartype does not do deep validation on dicts, so we do non-nested validation
+        if isinstance(data, dict):
+            for key, val in data.items():
+                if not isinstance(key, (str, int)):
```

Comment:
Is it necessary to check if all keys have the same type, i.e. all `str`, or all `int`, not mixed?

Reply:
Good question. You're right, this was designed to pass the unit test.